### ISSUE TITLE
Investigate options for handling aggregate species names from iNaturalist

### ISSUE THREAD

**Issue by kitenetter** [BiologicalRecordsCentre/iRecord#1222]
Taxa that identified as aggregates on iNat are being shown as segregates on iRecord, due to iNat handling aggregates differently to UKSI. One example:
- Recorded as "Complex Mesapamea secalis" on iNat: https://www.inaturalist.org/observations/90773026
- Displayed as "Mesapamea secalis" on iRecord: https://irecord.org.uk/record-details?occurrence_id=22872087

Notes from John:
"The record example for Complex Mesapamea secalis is actually reported by the API as taxon name Mesapamea secalis, with rank set to Complex. It doesn’t have agg. at the end of the name which is the UKSI standard, nor does it have Complex in the name. We don’t check the rank as part of the name. If we check the rank, then we’ll need a mapping table to map Complex (which doesn’t exist in UKSI) to agg, and perhaps automatically tolerate the absence of agg. at the end of species names. But the whole thing of mapping correctly from the iNat taxonomy to UKSI is a whole can of worms so there might be quite a lot to do once we look into all the issues."

We need to find a resolution to this for verifiers. Options:
- Spend time developing a dictionary mapping process as John describes
- Enable verifiers to do bulk redeterminations - see #674 
- Make it more obvious in iRecord that the iNat taxon is a "Complex", e.g. by adding an icon or other highlighting mechanism

**Comment by kitenetter** [comment]
Related to this is the issue of mismatches between iNat and UKSI taxonomy, e.g. the robberfly that is Machimus atricapillus in the UKSI is called Tolmerus atricapillus in iNat, and consequently the records never reach iRecord. A dictionary mapping table may allow both the name issue and the aggregate issue to be addressed.

**Comment by johnvanbreda** [comment]
@kitenetter I agree - some sort of dictionary mapping table is needed to resolve these issues.

**Comment by kitenetter** [comment]
@robin-hutchinson I think the starting point here is to get a list of the iNat taxon names that have been recorded in the UK, and then do a comparison to see which ones have no match in the UKSI. That should be straightforward for synonyms, but may be harder for aggregates.

If you can see a way of extracting a UK recorded taxon list from iNat itself that would be fine, otherwise please can you contact Sophie Ratcliff at NBN and ask if she can provide a table of records per taxon from their iNat download.

**Comment by robin-hutchinson** [comment]
The full list of species recorded in the UK is available to download from the checklist pages: https://www.inaturalist.org/check_lists/7190-United-Kingdom-Check-List?view=plain. I am just downloading it to try to comparison now.

**Comment by robin-hutchinson** [comment]
[Aggregate Matches.csv](https://github.com/BiologicalRecordsCentre/iRecord/files/10581078/Aggregate.Matches.csv)
[iNat UKSI Edited Match.csv](https://github.com/BiologicalRecordsCentre/iRecord/files/10581495/iNat.UKSI.Edited.Match.csv)
[iNat UKSI No Match.csv](https://github.com/BiologicalRecordsCentre/iRecord/files/10581496/iNat.UKSI.No.Match.csv)

I've looked at the iNat tables and found the aggregates that have fulfill the requirements to be shared with iRecord (research grade and CC-BY, CC0 or CC-BY-NC). I pulled out everything with a rank of "complex" and found the aggregate match if there is one - if there was not a match, I gave the species name and then listed the species within the iNat complex if this was under ten species. There are some cases where we might be able to record the species within the UK, as the other species within the complex are not present here (e.g. Austropotamobius pallipes complex).

I also exported a list of all of the iNat ids that exist within iRecord (regardless of verification status) to filter out names that are successfully moving to iRecord, and then tried to match them using the UKSI code. I've attached the taxa that have a match if the names are slightly edited (e.g. Deroceras invadens -> Deroceras (Deroceras) invadens, Amanita excelsa excelsa -> Amanita excelsa var. excelsa, Choerades marginata -> Choerades marginatus, Bothria subalpina -> Botria subalpina, Actenicerus siaelandicus -> Actenicerus sjaelandicus) or where the subspecies wasn't known, so it matched to the species instead (e.g. Rubus idaeus idaeus -> Rubus idaeus). It uses fuzzy matching up to two letters different to deal with spelling differences - this works when it is a list that I check for errors before using, but wouldn't work for the iNat-iRecord link as there isn't a manual checking stage. We might be able to resolve some of these issues using a similar process to the UKSI name-matching process in Indicia (https://uksi-sandbox.nhm.ac.uk/taxonmatch.php):

exact - there is an exact match on the name & authority with a NULL attribute
name - there is an exact match on the name only
subgenus - there is a match by either removing a bracketed subgenus in your name OR by allowing for a bracketed subgenus in your name
gender - there is a match by looking for a different gender ending to the species epithet. We do this by chopping off 2 characters from the end of your name and adding a wildcard
none - no match was found

There is also a spreadsheet for the taxa that there wasn't a match for. I will go through this final spreadsheet to separate out species with unknown synonyms (e.g. Pseudochorthippus parallelus) from species that are not in iRecord, and will pull out the incorrect matches from the edited spreadsheet to see if I can find a better match for those as well.

We could also incorporate the rank columns in the UKSI and Indicia to make sure that we only try to match on the same rank level? This would help to distinguish between the two Bombus lucorum entries in the UKSI (species and sensu lato, with different TVKs).

**Comment by johnvanbreda** [comment]
Thanks @robin-hutchinson. I think that it probably is a useful thing to restrict matches on rank, as there are a few cases of duplicate names across different ranks that I've come across. 

From my perspective there are 2 tasks that need to come out of this issue - firstly to get a list of mappings from iNat names to UKSI names (or TVKs/organism keys) that is acceptable to the community. Your proposed name matching rules sound good from my perspective but I'm probably not the right person to decide what is acceptable. The 2nd task will be to then integrate this list into the existing iNat importer and use it to ensure that we are able to import all records rather than a sub-selection. It may then be necessary to re-import the data (at least unverified records) to ensure we pick up the missing stuff.

For the Saint Helena iNat link, I have been thinking that I should add code to the sync which catches unmapped names and adds them to a sandbox list, so the records are still importable and the sandbox list can be merged into the main list later. This might be useful here after the work on the mappings is done, as the iNat taxonomy may change in future so we'd want to be able to resolve future mapping requirements.

**Comment by robin-hutchinson** [comment]
Thanks, I'll keep working on suggested mappings. Looking at this today, I found that https://jumear.github.io/stirfry/iNatAPIv1_taxa?q=chorthippus&is_active=false&per_page=500&rank=species gives synonyms - I haven't yet found a way to search the taxon tables on synonym id (synonym_id=501625 doesn't produce the correct result) but if I figured out how to do this, could we add code to search for the synonyms in the UKSI if a match for the recommended name isn't found, or download and use a dictionary table for junior synonyms? I can only download 10000 at a time which doesn't include everything.

**Comment by johnvanbreda** [comment]
If you can work out a mapping from the accepted taxon name on iNat to a list of the iNat synonyms then we could use that to attempt matches for records that fail on the first attempt. But, as the iNat API for an observation does not include the synonyms for the record (at least not as far as I can see from the examples), we would have to map from the iNat accepted name to the iNat synonym then see if that matches one of our names. That seems like an extra complexity vs just having a simple mapping table that lists iNat names to the UKSI equivalent.

**Comment by robin-hutchinson** [comment]
Thanks John - I think you are right - I was just worried about iNaturalist updating in future without us being aware so the new iNat to UKSI would not be included, but this would be covered by the sandbox list that we can periodically check for missing synonyms?

**Comment by johnvanbreda** [comment]
Yes, there is already a log table that lists the failures. We can easily extract the list of non-matched names from this log.

**Comment by robin-hutchinson** [comment]
Brilliant, thank you, that will be perfect then!

**Comment by kitenetter** [comment]
@robin-hutchinson and @kitenetter to review taxon-match table following update of UKSI within Indicia, then provide table to John.

**Comment by kitenetter** [comment]
@robin-hutchinson can you run a check against the updated Indicia table and let me see the results.

**Comment by robin-hutchinson** [comment]
No problem - the code is running now, I'll email you the generated files

**Comment by robin-hutchinson** [comment]
Of course - I've got that column now, I'll just email them over.

**Comment by kitenetter** [comment]
@robin-hutchinson this version takes in the plant names that @Sam-Amy matched, plus a few more from the UKSI matching tool and other detective work. We  now have matches for 163 of the 851 unmatched list, including almost all for taxa with over 100 records. Most of the missing names have fewer than 10 records each.

A few matched names appear to be identical to the iNat originals so I'm not sure why they weren't being recognised on import to iRecord.

Suggest we go with the 163 names that we can deal with, and some of the others may get picked up in future UKSI updates.

[iNat UKSI No Match_MH.xlsx](https://github.com/BiologicalRecordsCentre/iRecord/files/12323046/iNat.UKSI.No.Match_MH.xlsx)

**Comment by johnvanbreda** [comment]
Is this ready for me to work on setting up the matching?

**Comment by kitenetter** [comment]
@robin-hutchinson can you combine the 163 names from the latest spreadsheet into your longer list of matched names and pass that on to @johnvanbreda

**Comment by robin-hutchinson** [comment]
All done - I'll reassign this to John

**Comment by johnvanbreda** [comment]
I'm having a few issues with the mappings table @robin-hutchinson. What I need to map each row to in the database is a single taxon name on the UKSI list, ideally I dont want to just use the preferred name for the rTVK as that loses detail on what name was actually used. I have a name_mapped field so I can combine that with the rTVK field to select a name from UKSI to use, but I sometimes get duplicates mostly due to variations in authority strings, but I can get round that by just picking one and preferring names that are accepted. But I also have names that don't match at all - e.g. for _Taraxacum officinale_ the mapped name is given as "Taraxacum officinale F.H.Wigg. s.s." but we only have "Taraxacum officinale agg." with or without an authority of Weber. There are approx. 100 names that don't easily match up.

Is it possible to add a TVK for the mapped name to the spreadsheet? That would make the mapping much simpler.

**Comment by robin-hutchinson** [comment]
Hi John - sure I will work on that and post the updated table here, thanks!

**Comment by Sam-Amy** [comment]
That sounds suspiciously like the number of names I matched using the BSBI DDb taxon parser, so perhaps it doesn't work quite as I thought. I will have a look too @robin-hutchinson if you could give me a copy of the final spreadsheet please?

**Comment by kitenetter** [comment]
@robin-hutchinson can you check whether the translation table includes iNat's "Helix lucorum", which needs to be matched to the UKSI "[Helix (Helix) lucorum](https://uksi-sandbox.nhm.ac.uk/taxon.php?linkKey=NHMSYS0020978827)". (We've had an enquiry from Conch Soc.)

**Comment by kitenetter** [comment]
@robin-hutchinson another query relating to Conch Soc - please can you check that the translation table is handling slug names correctly as detailed in #1154

**Comment by johnvanbreda** [comment]
@robin-hutchinson @kitenetter is there a finalised spreadsheet of mappings available for this?

**Comment by johnvanbreda** [comment]
Have now added code so once the mappings are uploaded into the `rest_api_sync_taxon_mappings` table, the previous failures can be rescanned. Info on how to do this in the rest_api_sync module example config file.

**Comment by kitenetter** [comment]
@robin-hutchinson can we get the translation table finalised for this?

There are now three requests from Conch Soc that we need to address:
- The one [described above](https://github.com/BiologicalRecordsCentre/iRecord/issues/1222#issuecomment-1777591178) for Helix lucorum
- The [Arion subgenus issue](https://github.com/BiologicalRecordsCentre/iRecord/issues/1154)
- A new request: taxa entered as "Hermaea bifida" on iNat need to be matched to "Hermaea bifida sensu stricto"

**Comment by robin-hutchinson** [comment]
Hi Martin,

I have the translation table for the "fuzzy matches", which includes the Helix lucorum issue. I don't think we can resolve the Arion subgenus issue this way as the taxon name that appears for us is just "Arion" ("ater-group Arion Slugs" is the common name), unless we can see the common names from iNaturalist?

[iNat UKSI matches.csv](https://github.com/user-attachments/files/23058573/iNat.UKSI.matches.csv)

I've just included the iNat name and the matching tvk to make it lightweight (to be stored), but can add extra columns if that would be useful, these two columns would stay the same.

What I don't think is working here is the complexes - which correctly/incorrectly match to the species level det depending on the rank. I've added those as a separate table below (the complex matches), based on whether we have an equivalent aggregate or not.

[iNat complexes.csv](https://github.com/user-attachments/files/23058728/iNat.complexes.csv)

**Comment by sacrevert** [comment]
For the iNat complexes list, Galium mollugo (Galium album in UK), Galium aparine, Symphytum officinale and Quercus ilex could just be mapped to the corresponding species in the UKSI. The confusion species within the iNat complexes are not recognised/known to occur in Britain anyway.

Also, in the iNat UKSI matches.csv file
- Thymus praecox brittanicus would be better matched to Thymus drucei (https://uksi-sandbox-test.nhm.ac.uk/taxon.php?linkKey=NHMSYS0000464446)
- Sanguisorba minor balearica would be better matched to https://uksi-sandbox-test.nhm.ac.uk/taxon.php?linkKey=NHMSYS0021127626
- Dactylorhiza praetermissa schoenophila would be better matched to https://uksi-sandbox-test.nhm.ac.uk/taxon.php?linkKey=NHMSYS0021874032

**Comment by johnvanbreda** [comment]
@robin-hutchinson is the matches CSV file you posted on Oct 22 the final one, or do you need to do some additional work regarding complexes?

**Comment by kitenetter** [comment]
@robin-hutchinson we need to pick up on this again now that the UKSI update has happened, let me know if there are still things we need to resolve.

**Comment by robin-hutchinson** [comment]
I've just updated the two taxonomy tables with the new UKSI version and an extra year of iNaturalist data, which has removed some rows where we now have a complete match and added more that weren't in use in iNaturalist previously. I've tagged the new species in both, Martin and Oli could you check if there are missing matches in the new rows, please?

[iNat UKSI matches 20260807.csv](https://github.com/user-attachments/files/30862997/iNat.UKSI.matches.20260807.csv)

[iNat complexes 20260807.csv](https://github.com/user-attachments/files/30863000/iNat.complexes.20260807.csv)

**Comment by sacrevert** [comment]
thanks @robin-hutchinson 
If the correct filter to look at is `present_2025 = yes`, then these all look fine for the groups I can judge.

**Comment by kitenetter** [comment]
**UKSI matches**
Some of iNat's subspecies are being mapped to UKSI species, while others are being mapped to UKSI subspecies. would be good to document how/why these choices have been made.

**Complexes**
A few iNat complexes have been matched to UKSI segregate species, and again it would be good to know why. E.g. @sacrevert are you happy that iNat's "complex" version of Galium aparine, Galium mollugo, Quercus ilex and Symphytum officinale are being matched to segregate species in UKSI?

And there are quite a lot of iNat complexes that do not have UKSI equivalents. @sacrevert do you see any benefit in mapping these to the relevant genus, or do you think they are best left so that do not get imported to iRecord? For example, we could map "Complex Solanum nigrum" to genus Solanum.

**Comment by sacrevert** [comment]
For plants, the UKSI issue looks ok because those cases of subspecies mapping to species are those where we have no other recorded subspecies and so we don't have the nominate subspecies separately defined in the UKSI currently.

For the complexes we don't currently accept whatever global taxonomy is splitting these, so the maps are currently appropriate.

Both these conclusions could change in the future of course.

**Comment by kitenetter** [comment]
Okay, I think it was some of the birds that I looked at where the UKSI does have an equivalent subspecies, but we are proposing to map to species. That my be the most pragmatic solution in a UK context.

A lot of the iNat complexes appear to be 'artificial' aggregates that have been created in order to enable recording of similar species that are hard to tell apart, rather than being splits of accepted species concepts. E.g. the [Solanum nigrum complex](https://uk.inaturalist.org/taxa/1351686-Solanum-nigrum#taxonomy-tab) contains a wide range of species, some of which are in UKSI and some presumably are not known to occur in the UK. Arguably it might be appropriate to bring these into iRecord at genus level so that verifiers could redetermine within the UK context if they wished, but I can see pros and cons to that approach.

**Comment by robin-hutchinson** [comment]
For the UKSI subspecies - I tried to match these at the subspecies level first using partial matches, if I couldn't find a match I removed the subspecies and matched at the species level. Could you let me know the birds that this isn't working for?

We can map the aggregate to the genus (I made the matches by looking for aggregates in the same genus as the complex name, and then checked if the matches were appropriate), happy to do this if it works for the verifiers.

**Comment by sacrevert** [comment]
>A lot of the iNat complexes appear to be 'artificial' aggregates that have been created in order to enable recording of similar species that are hard to tell apart, rather than being splits of accepted species concepts. E.g. the [Solanum nigrum complex](https://uk.inaturalist.org/taxa/1351686-Solanum-nigrum#taxonomy-tab) contains a wide range of species, some of which are in UKSI and some presumably are not known to occur in the UK. Arguably it might be appropriate to bring these into iRecord at genus level so that verifiers could redetermine within the UK context if they wished, but I can see pros and cons to that approach.

I think all that will happen if we do this is that they will be completely ignored by verifiers for the most part; I think the information loss will be lower if they are matched to the most probable concept as currently and then verifiers are free to redetermine if they want to/can.

**Comment by kitenetter** [comment]
@robin-hutchinson here are just a few examples of the bird names:

[iNat.UKSI.matches.20260807_bird examples.xlsx](https://github.com/user-attachments/files/30939126/iNat.UKSI.matches.20260807_bird.examples.xlsx)

It's a tricky one as it may well be that some recording schemes would prefer not to use the subspecies names in a UK context. But we probably do need to be as consistent as we can when transferring the names from iNat.

@sacrevert you say "matched to the most probable concept as currently" but my understanding is that a lot of names on the list of complexes will not be matched at all (e.g. the Solanum example), and will therefore not appear in iRecord. Which may be the best outcome if UKSI does not have a comparable taxon to match them to.

**Comment by sacrevert** [comment]
@kitenetter ok, I was filtering on family groups so i missed the unmatched plant concepts. ~~I don't have a strong opinion given how little information this represents for plants and the time required to resolve it properly.~~
Actually, having looked again and seen Leucanthemum vulgare on that list, I do have an opinion, which is just that we should naively attach the data to the implied current species concept in the UKSI and then let recorders assess them as they will.
FURTHER EDIT: also, if this approach is taken and then in the future our taxon is recognised as containing other species-level taxa, then presumably the best thing is to rename the same concept as some sort of agg. in the UKSI with the nested sensu stricto taxon created as a new concept (organism concept). This would mean that a current match to our species as currently understood remained correct in terms of the information content of the record.

**Comment by robin-hutchinson** [comment]
I've just updated the matches to remove the subspecies that are in the UKSI.

For complexes: 

We could create an aggregate as the junior synonym of the species name to match these (to preserve the original record but point at the species as we currently have it). This would have to be in the UKSI (and would therefore wait for the next update from the UKSI to Indicia). We might in that case not want to import those complexes for the time being, as they could be verified and locked at the species level (so we couldn't retroactively point the iNat record to the complex). 

Or - from your further edit, did you want the records in Indicia as it currently stands to also be changed to the aggregate if its added, and the species name to become a separate entity to be used for future recording (with a new TVK and Organism Key)? In that case, we could just remove the species you want to apply that to from the spreadsheet altogether, as complexes are already being matched to the species name (which we need to avoid for the complexes in the spreadsheet).

[iNat UKSI matches 20260828.csv](https://github.com/user-attachments/files/31560182/iNat.UKSI.matches.20260828.csv)

**Comment by sacrevert** [comment]
Thanks @robin-hutchinson In that file I've spotted one vascular plant error:
Clinopodium menthifolium adscendens should map to Clinopodium adscendens, not C. menthifolium

**Comment by sacrevert** [comment]
@robin-hutchinson Regarding the other question, I was really referring to scheme-level management of the UKSI. I don't think we should be doing anything to the UKSI unless we are recognised as the primary managers for a given taxon group. When I was referring to UKSI edits, I was thinking about future changes arising from scheme-level decision making. Cheers

**Comment by kitenetter** [comment]
@robin-hutchinson if you can correct the entry for Clinopodium that Oli highlighted I think we can then pass the csv files to John (and assign this issue to him).

**Comment by robin-hutchinson** [comment]
No problem - all done! [iNat UKSI matches 20260911.csv](https://github.com/user-attachments/files/32100876/iNat.UKSI.matches.20260911.csv)

**Comment by kitenetter** [comment]
@robin-hutchinson can you confirm which csv files contain the full and final list of taxa to be mapped. Is it these two:
- https://github.com/user-attachments/files/32100876/iNat.UKSI.matches.20260911.csv
- https://github.com/user-attachments/files/30863000/iNat.complexes.20260807.csv

**Comment by johnvanbreda** [comment]
Presumably I should do the matches CSV file, plus the complexes CSV file where a UKSI name is given, but not for the other rows where no link to UKSI is given?

I've added some tweaks to the code for v9.27 ready for this, for example I have noticed that crosses using the unicode × character don't match our UKSI names which use a normal x, but I can automatically handle that. 

One thing I've noted - for some names, we only know the scientific name from iNaturalist, but we have 2 different organisms in UKSI with the same scientific name, e.g. with different author strings. An example is _Solanum nigrum_ where there is an organism with author=L. and another where the author is "auct. non L". I can't resolve these automatically, perhaps we should add them to the mappings? I think it would be a good idea to get the initial mappings set up first, then review to see what names are still being blocked as they can't be mapped.

**Comment by kitenetter** [comment]
Agree that if we can do a further review of what names are **not** being matched that would be good, but also I'm keen to start matching the ones we can do as soon as this is possible. It's likely that there will always be a proportion of names that can't easily be matched or for which the correct match is unclear.

**Comment by sacrevert** [comment]
Just to say that the Solanum nigrum example may be quite rare. Solanum nigrum auct. non L. is a name that probably applies to two other valid taxa (S. nitidibaccatum and S. sarrachoides), it is just that the UKSI doesn't seem to capture this at the current time for some reason (not sure whether @japonicus can suggest a solution for this to Chris Raper?)

**Comment by johnvanbreda** [comment]
There may also be issues with sensu lato versions of names as well, as we don't know which to match to. As @kitenetter says it makes sense to proceed with the initial fixes then review - we keep track of all the dropped records so they can be imported again once sorted anyway.
@robin-hutchinson just waiting on confirmation if I should include the complexes file for the names that match?

**Comment by japonicus** [comment]
> One thing I've noted - for some names, we only know the scientific name from iNaturalist, but we have 2 different organisms in UKSI with the same scientific name, e.g. with different author strings. An example is _Solanum nigrum_ where there is an organism with author=L. and another where the author is "auct. non L". I can't resolve these automatically, perhaps we should add them to the mappings?

The plant names where the authority is prefixed with _auct. non_ are junk concepts that shouldn't be matched against. They'll ether be ambiguous split concepts (e.g. the Solanum nigrum example) or names that have been misused for recording at some point, where the accepted concept is unknown.

They should probably be marked as deprecated in the UKSI. In last year's taxon update we didn't add any new 'auct.' names and tried to tidy up some, treating them as junior synonyms, but others got missed or couldn't be resolved.

**Comment by sacrevert** [comment]
Thanks @japonicus -- are you able to ask Chris to deprecate the relevant concepts in the UKSI, I assume this means that they will be then unavailable for data entry in iRecord and will also help simplify these iNaturalist matching issues in future

**Comment by japonicus** [comment]
@sacrevert I'll discuss this with Chris. Hopefully it should be straightforward to supress the _auct._ names, but I don't know what the implications are for occurrences tied to deprecated names.

**Comment by johnvanbreda** [comment]
Thanks - I can modify our importer to ignore the auct. non names then. If a name gets deprecated in UKSI, it still exists in our data so existing occurrences aren't affected - it's just the lookup used for searching names when recording that is affected.
